src/TorchSharp/JIT/ScriptModule.cs

```diff
@@ -0,0 +1,309 @@
+                                    break;
+                                }
+
+                            case Tensor tensor when (deviceType != tensor.device_type || deviceIndex != tensor.device_index): {
+                                    var t = tensor.to(deviceType, deviceIndex);
+                                    field.SetValue(this, t);
+                                    ConditionallyRegisterBuffer(fieldName, t);
+                                    break;
+                                }
+                            }
+                        }
+
+                        _deviceType = deviceType;
+                        _deviceIndex = deviceIndex;
+                    }
+
+                    Debug.Assert(_deviceType == DeviceType.CUDA || _deviceIndex == -1);
+
+                    return this;
+                }
+
+                private DeviceType _deviceType = DeviceType.CPU;
+                private int _deviceIndex = -1;
+
+                /// <summary>
+                /// Convert the parameters and buffers.
+                /// </summary>
+                /// <returns></returns>
+                public override nn.Module to(ScalarType dtype)
+                {
+                    THSJIT_Module_to_dtype(handle, (sbyte)dtype);
+                    CheckForErrors();
+
+                    foreach (var (_, sm) in named_children()) sm.to(dtype);
+                    foreach (var field in GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)) {
+
+                        var fieldName = field.Name;
+                        var value = field.GetValue(this);
+
+                        switch (value) {
+                        // This test must come before the Tensor test
+                        case Modules.Parameter param when dtype == param.dtype:
+                            continue;
+
+                        case Modules.Parameter param: {
+                                var t = param.to(dtype);
+                                t.retain_grad();
+                                var p = new Modules.Parameter(t, param.requires_grad);
+                                field.SetValue(this, p);
+                                ConditionallyRegisterParameter(fieldName, p);
+                                break;
+                            }
+
+                        case Tensor tensor when dtype == tensor.dtype:
+                            continue;
+
+                        case Tensor tensor: {
+                                var t = tensor.to(dtype);
+                                field.SetValue(this, t);
+                                ConditionallyRegisterBuffer(fieldName, t);
+                                break;
+                            }
+                        }
+                    }
+
+                    return this;
+                }
+
+#if false   // These functions "work," but the native code doesn't seem to find any interesting information.
```

Comment:
>if [](http://example.com/codeflow?start=1&length=2)

we'll keep this code for now? what is the next step here?

Reply:
I want to try to figure out why the scripts don't seem to contain any information on element type or dimensions.